### PR TITLE
Fix PARTITION BY

### PR DESCRIPTION
**Pull request by mandrews**
[See Also](https://github.com/socialcast/masamune/pull/176/files)

### CODE REVIEW

File: lib/masamune/transform/postgres/snapshot_dimension.psql.erb [hunk 1]
```diff
@@ -28,11 +28,11 @@ http://www.postgresql.org/message-id/CAGnEbohhKmW55oB0FpQd3naXBkwi74E=8DRZBGFjS-
 WITH ranges AS (
   SELECT *,
   CASE WHEN delta = 0
-  THEN 1 ELSE NULL END r
+  THEN 1 ELSE 0 END r
```
Comment (mandrews) [inline lib/masamune/transform/postgres/snapshot_dimension.psql.erb]: The CTE `ranges` is used to collect delta records together with their associated complete record denoted by `window_id`
This code assumes that a complete record will always be available before a delta record.  If this assumption is violated, this results in a `NULL` `window_id` for delta records that happen before complete records

File: lib/masamune/transform/postgres/snapshot_dimension.psql.erb [hunk 2]
```diff
@@ -28,11 +28,11 @@ http://www.postgresql.org/message-id/CAGnEbohhKmW55oB0FpQd3naXBkwi74E=8DRZBGFjS-
...
   FROM <%= source.name %>
 ), windows AS (
   SELECT *,
-  SUM(r) OVER (ORDER BY <%= target.window("start_at #{order}", 'delta', 'source_uuid DESC').join(', ') %>) window_id
+  SUM(r) OVER (PARTITION BY <%= target.window.join(', ') %> ORDER BY <%= ["start_at #{order}", 'delta', "source_uuid #{order}"].join(', ') %>) AS window_id
```
Comment (mandrews) [inline lib/masamune/transform/postgres/snapshot_dimension.psql.erb]: complete ordering works if all delta records happen after a complete record, but ideally the `window_id` should be computed independently for each `PARTITION `

File: lib/masamune/transform/postgres/snapshot_dimension.psql.erb [hunk 3]
```diff
@@ -41,15 +41,15 @@ WITH ranges AS (
...
       <%- target.insert_values(window: 'w').each do |value, last| -%>
       <%= value %><%= ',' %>
       <%- end -%>
       start_at AS start_at
     FROM
       windows
-    WINDOW w AS (PARTITION BY <%= target.window('window_id').join(', ') %> ORDER BY start_at <%= order %>)
-    ORDER BY <%= target.window("start_at #{order}", "id DESC", 'window_id').join(', ') %>
+    WINDOW w AS (PARTITION BY <%= target.window('window_id').join(', ') %> ORDER BY <%= ["start_at #{order}", "source_uuid #{order}"].join(', ') %>)
+    ORDER BY <%= target.window("start_at #{order}", "source_uuid DESC", 'window_id').join(', ') %>
```
Comment (mandrews) [inline lib/masamune/transform/postgres/snapshot_dimension.psql.erb]: The `SELECT DISTINCT` is actually de-duplicating records that occur simultaneously, so `ORDER BY source_uuid DESC` is necessary to pick the latest source record when resolving conflicts

File: lib/masamune/transform/postgres/snapshot_dimension.psql.erb [hunk 4]
```diff
@@ -41,15 +41,15 @@ WITH ranges AS (
     <%- end -%>
     consolidated.start_at
   FROM (
-    SELECT DISTINCT ON (<%= target.window('start_at', 'id').join(', ') %>)
```
Comment (mandrews) [inline lib/masamune/transform/postgres/snapshot_dimension.psql.erb]: not sure why `id` is in this `SELECT DISTINCT` clause- seems vestigial and defeats the purpose

**Comment by nksocast** [comment]
looks good :+1:

**Pull request merged**